### DISCUSSION TITLE
How to model a mobile manipulator?

### DISCUSSION THREAD

**Discussion by LHT-Liu** [stack-of-tasks/pinocchio#2184] [Q&A] [answered]
i want to know how to modeling the mobile manipulator with float-base,as shown below

<img width="595" alt="image" src="https://github.com/stack-of-tasks/pinocchio/assets/64388319/98c28209-3e94-4abe-a714-18efd44b7edd">

I can model the manipulator with traditional method,but i don't konw how to model the mobile . i would like to learn how pinocchio handles this kind of problem？

**Comment by jcarpent** [comment, marked as answer]
FOr the mobile platform, you can use PlanarJoint to work with X and Y translations and one rotation around the vertical axis.
Best,
Justin

**Reply by LHT-Liu** [reply]
thanks for your reply.
and I have a question how do PlanarJoint and urdf file correspond？do I get correct model form urdf that mobile manipulator(mobile platform and manipulator )？
Is there any similar example in pinocchio?

**Reply by stephane-caron** [reply]
In the "type" attribute of the joint element there is a "planar" option, see [urdf/XML/joint](https://wiki.ros.org/urdf/XML/joint). Those will be loaded as planar joints in Pinocchio.

**Reply by LHT-Liu** [reply]
@stephane-caron thanks for your reply, i understand what you said. Is there any similar example（mobile manipulator） in pinocchio?

**Reply by LHT-Liu** [reply]
and，this considers the mobile platform as a whole. Have you directly considered the situation of four wheels?

**Reply by stephane-caron** [reply]
I don't know of examples in Pinocchio itself, but there are examples in libraries that rely on Pinocchio such as:

- [PLACO](https://github.com/Rhoban/placo): where there is a [`WheelTask`](https://placo.readthedocs.io/en/latest/kinematics/wheel_task.html) suited to wheels, which can optionally be omnidirectional

https://github.com/stack-of-tasks/pinocchio/assets/1189580/137b665e-6a93-48e8-ad5f-be2f7bb9340e

- [Pink](https://github.com/stephane-caron/pink): where there are examples [for the Stretch RE1](https://github.com/stephane-caron/pink/tree/main/examples#mobile-stretch) mobile-base manipulator (disclaimer: I'm a maintainer of Pink) but no wheel task yet

https://github.com/stephane-caron/pink/assets/1189580/711c4b92-6234-41bd-945b-e6c043f6b2e6

**Reply by LHT-Liu** [reply]
Thanks, Let me examine the example you provided

**Comment by LHT-Liu** [comment]
> i want to know how to modeling the mobile manipulator with float-base,as shown below
> 
> <img alt="image" width="595" src="https://private-user-images.githubusercontent.com/64388319/316209414-98c28209-3e94-4abe-a714-18efd44b7edd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1ODEwNzAsIm5iZiI6MTcxMjU4MDc3MCwicGF0aCI6Ii82NDM4ODMxOS8zMTYyMDk0MTQtOThjMjgyMDktM2U5NC00YWJlLWE3MTQtMThlZmQ0NGI3ZWRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDEyNTI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyMzUyOTdiZTY4YjMyMmE2ZDNjZGM0ZGEyOGU5NGRmNWNjMDY2YjAxZWUwOWJmYzA3NDQ2MjVhMDNlODQ1YjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SUbTKByEOjt62jK3Ea2AESE9_9o42Lpkg9sYaWhGCzc">
> I can model the manipulator with traditional method,but i don't konw how to model the mobile . i would like to learn how pinocchio handles this kind of problem？

@stephane-caron hello，i have a question, this is my urdf,
```
<link name="world"/>

<link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-25428000000000 -1005100 -2223600000000"/>
      <mass value="110"/>
      <inertia ixx="9.1864E+40" ixy="-4.7485E+35" ixz="-1.0505E+42" iyy="1.2105E+43" iyz="-4.1525E+34" izz="1.2013E+43"/>
    </inertial>

    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="xxxxxxxxx/base_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>

    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="xxxxxxxxxxx/base_link_sc.STL"/>
      </geometry>
    </collision>

  </link>

  <joint name="base_link_planar_joint" type="planar">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="world"/>
    <child link="base_link"/>
  </joint>

  `other links, joints`
```

i can get `q0 = ` by `pin.neutral(model)`,
<img width="466" alt="image" src="https://github.com/stack-of-tasks/pinocchio/assets/64388319/a3df46fd-9b6d-4ffb-b2b6-21a9dde65efd">

but i noticed that I set q0 = np.array([0, 0, 1, 0, xxxx]). The first two are to set the values ​​of x and y, but what does 1 mean? What do the third and fourth values ​​mean? I feel It seems to be the scaling factor, not the rotation angle of the z-axis？

thank you very much

**Reply by stephane-caron** [reply]
Hmm, the documentation page related to that would be [Joint dynamics](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/md_doc_c-maths_b-joints.html), but the information you are looking for seems to be missing.

The configuration vector for planar joints is $(x, y, \cos \theta, \sin \theta)$ where $\theta$ is the angle (to be precise: of the rotation from the child frame to the parent frame). I opened a proposal to add this comment to the docs in https://github.com/stack-of-tasks/pinocchio/pull/2211

**Reply by LHT-Liu** [reply]
@stephane-caron  thanks，I never thought it was cos (theta), sin (theta), why not use one theta, but use two quantities?

**Reply by jorisv** [reply]
You can find some explanation [here](https://github.com/stack-of-tasks/pinocchio/issues/794#issuecomment-490050573) and [here](https://github.com/stack-of-tasks/pinocchio/issues/2120#issuecomment-1857908521)

**Reply by LHT-Liu** [reply]
Ok,thanks
